Profile registration service.

rushtong opened the pull request:
Orchestration side of profile registration service. 

Parent story (https://broadinstitute.atlassian.net/browse/DSDEEPB-1815) is not complete without the UI side which is not implemented yet.

davidangb commented:
Take a look at the validation approach I used in #123; I think it might be simpler. Specifically, look at the FireCloudPermission case class. I put the require() statements right in the case class, and made its vals Strings instead of Option[String]s. This means we validate at the time of creation of the case class, instead of allowing it to be created and validating after the fact.

Additionally, note that I added handleRejections(entityExtractionRejectionHandler) to the POST route in MethodsService; this gives a nice error message to the user if the POST payload can't be unmarshalled into the case class because of the validation.

(this is not a full-scale code review)

davidangb commented on src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileActor.scala:
```diff
@@ -0,0 +1,86 @@
+              context.parent ! RequestCompleteWithErrorReport(response.status, response.toString)
+              context stop self
+            case _ =>
+              context.parent ! RequestCompleteWithErrorReport(InternalServerError, "Failure posting profile: [%s]".format(profileProperties))
```
nitpick: all the profile keys were saved; it's just the isRegistrationComplete that failed. Update the log message so it's not confused with the one at line 80.

davidangb commented on src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileActor.scala:
```diff
@@ -0,0 +1,86 @@
+              context stop self
+          }
+
+        case Success(responses) =>
```
This block is also used in the entity-deletion code. It would be nice to extract this into a reusable utility or some Spray command (a route? a directive? I'm not even sure what the right object is). Don't hold up the PR for this; it's an optimization.

davidangb commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/FireCloudRequestBuilding.scala:
```diff
@@ -31,4 +31,13 @@ trait FireCloudRequestBuilding extends spray.httpx.RequestBuilding {
 
+  def dummyUserIdHeaders(userId: String) = {
+    addCredentials(OAuth2BearerToken("mF_9.B5f-4.1JqM")) ~>
+      addHeader(RawHeader("OIDC_CLAIM_sub", userId)) ~>
```
remove the _sub. I believe we've determined that the openid-connect proxy creates _sub, while the oauth proxy creates _user_id. Let's use just _user_id to mimic the current proxy.

davidangb commented on src/test/scala/org/broadinstitute/dsde/firecloud/service/UserServiceSpec.scala:
```diff
@@ -0,0 +1,153 @@
+            new RequestBuilder(method)(s"/profile/$exampleKey") ~>
+              dummyUserIdHeaders(uniqueId1) ~> sealRoute(routes) ~> check {
+              log.debug(s"$method /profile/$exampleKey: " + status)
+              status shouldNot equal(MethodNotAllowed)
```
here (and elsewhere), I'd rather see:
```
status should equal(OK)
```
... in other words, test for the specific positive case, not a single negative case.

rushtong replied:
I'm going to disagree here. If we test for OK, then what we're testing Thurloe logic, not orchestration logic. Where the call is a passthrough, the only thing orchestration is doing is to pass the call onto some downstream service. Orchestration doesn't know what the response will be nor does it care what the response is. I could see the point if orchestration did something with the response, but in these passthrough cases, it does not.

davidangb replied:
Hmmmm, good point, I agree with you after hearing your stance. All set.

rushtong commented:
FireCloudPermission is much cleaner - thanks for the pointer!

rushtong commented:
@davidangb - I was able to successfully test this through swagger by running both UI and Orch locally and it is now ready for re-review. 
Summary:
* PR comments addressed
* Moved profile validation to require statements in Profile
  * This allowed for cleaning up `UserService` quite a bit
  * Maybe we should move common validation methods (nonEmpty, etc.) to a common object? This seems like a thing we will need more of.
* Updated tests somewhat. In one non-passthrough case, I was mistakenly not checking for OK status - that is fixed.
* `fullName` is now just `name` - UI is sending over name, not fullName. 
* Added `MalformedRequestContentRejection` to `ErrorReport`. This seems to work correctly in all tests.

davidangb commented on src/main/scala/org/broadinstitute/dsde/firecloud/model/Profile.scala:
```diff
@@ -10,3 +8,28 @@ case class ThurloeKeyValue(
+}
+
+object ProfileValidator {
+  def nonEmpty(field: String): Boolean = {
```
couldn't this just be:
```
def nonEmpty(field: String) = !field.trim.isEmpty
``
?

I know the match/case is all Scala-like, but still ...

rushtong replied:
Why yes, yes it can - I like the one liners where appropriate.

davidangb commented:
One minor code comment inline; otherwise good to go, works for me locally.

Feel free to fix the comment, squash if desired, and merge ... or push up another rev if that's easier, and I'll click the merge button.

rushtong commented:
Got the OK to merge from David in hipchat.
